[WIP] Improve solution load performance

saul opened the pull request:
Fixes #2793, fixes #2916, fixes #1944.

This is a very WIP pull request that improves solution load performance by moving ComputeSourcesAndFlags from being a UI thread operation to a background thread operation.

The design consists of a new ProjectFlagsService which processes a queue of projects that are dirty and need their sources/flags recalculating. It takes an exclusive lock on VS's design-time build and then runs an asynchronous build of the project (stopping just before invoking the compiler).

This PR also contains changes to ProjectsSitesAndFiles options caching. These changes mean that the Dense solution loads in 20 seconds instead of hanging VS forever. The ProjectsSitesAndFiles changes brought the load time down to 2:50mins, and the off-loading of ComputeSourcesAndFlags brings it down further to 20 seconds.

### Remaining work

- [x] Investigate whether Roslyn workspace setup can now be made synchronous
- [ ] Revert changes conflicting with #3002 
- [ ] Use MEF for ProjectFlagsService
- [ ] Prioritise the currently viewed project when the active document is changed
- [ ] Be resilient to Path.GetFullPath failing in reference node properties
- [ ] Fix sources and flags results not being passed down to the background compiler

saul commented:
I could do with a hand with the following issues:

- I can't get the ProjectFlagsService registered with MEF. When I do GetService on the site service provider in the ProjectSystem, it cannot find the service.
- When InvalidateConfiguration is called on the background request when the sources and flags have been calculated by MSBuild, it doesn't seem to update IntelliSense (still squiggles everywhere)

vasily-kirichenko commented on vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs:
```diff
@@ -208,7 +209,7 @@ type
                              DefaultToNonHotURLs = true,
                              EnableCommenting = true,
-                             CodeSenseDelay = 100,
+                             CodeSenseDelay = 0,
```
👍

saul replied:
Need to actually investigate if it's made a difference...

forki commented:
is there already noticable difference?

saul commented:
@forki with regards to what?

majocha commented on vsintegration/src/FSharp.ProjectSystem.FSharp/ProjectFlagsService.fs:
```diff
@@ -0,0 +1,123 @@
+open Microsoft.Build.Execution
+
+(*
+[<Export(typeof<ProjectFlagsService>)>]
```
Just [<Export>] would do?

vasily-kirichenko replied:
Maybe make it a Workspace Service instead like this http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Shared/DefaultDocumentSupportsSuggestionService.cs?

vasily-kirichenko replied:
...then query it like this http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/CommandHandlers/GoToAdjacentMemberCommandHandler.cs,53

saul replied:
Under the hood it still just uses ExportAttribute though - so I doubt that would help. Thanks for the input though :)

majocha commented on vsintegration/src/FSharp.ProjectSystem.FSharp/ProjectFlagsService.fs:
```diff
@@ -0,0 +1,123 @@
+type internal ProjectFlagsService 
+    [<ImportingConstructor>]
+    (
+        [<Import(typeof<SVsBuildManagerAccessor>)>] accessor : IVsBuildManagerAccessor
```
I'm not sure you'll get this from MEF.

IServiceProvider GetService maybe?

saul replied:
Thanks - it looks like I need to call GetService on IComponentModel instead of on this.Site 👍

majocha replied:
I experimented a bit and it turns out exports from other assemblies like FSharp.Editor work fine and are resolved with IComponentModel. Anything declared here fails.

saul replied:
That's bizarre, thanks for helping me with this. @cartermp do you know anyone who could weigh in/explain why our project system assembly doesn't seem to be linked up to MEF?

cartermp replied:
I have no idea.  @dsyme would you know, or know who might know?

dsyme replied:
@saul @cartermp I'd ask @OmarTawfik, @brettfo, @KevinRansom .  It may be just historical. Note I know almost nothing about MEF

majocha replied:
@saul might be helpful, too [https://github.com/Microsoft/vs-mef](https://github.com/Microsoft/vs-mef)

vasily-kirichenko commented on vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs:
```diff
@@ -1558,7 +1608,7 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
 #endif                        
                         x.ComputeSourcesAndFlags()
-                        if not(projectSite.State = ProjectSiteOptionLifetimeState.Opened) then
+                        if projectSite.State <> ProjectSiteOptionLifetimeState.Opened then
```
👍 :)

majocha commented on vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs:
```diff
@@ -316,29 +328,50 @@ and
+            |> Seq.map(fun ref -> (getProject ref.ProjectId).FilePath)
+            |> hashSetIgnoreCase
+
+        for referencedSite in  do
```
`updatedProjects.Except(workspaceProjects)` ?

saul replied:
Yes I was wondering that... will change them to use that tomorrow 👍

forki commented:
@saul I guess my question is if this work already shows performance benefits for your projects.

saul commented:
@forki as in the PR description:

> These changes mean that the Dense solution loads in 20 seconds instead of hanging VS forever. The ProjectsSitesAndFiles changes brought the load time down to 2:50mins, and the off-loading of ComputeSourcesAndFlags brings it down further to 20 seconds.

So yes, in large solutions with dense dependency graphs, it's essentially gone from impossible to load to 20 seconds 😃

saul commented on vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs:
```diff
@@ -47,19 +47,20 @@ type internal FSharpCheckerProvider
+                            let solution = workspace.CurrentSolution
+                            let documentIds = solution.GetDocumentIdsWithFilePath(fileName)
+                            if not documentIds.IsEmpty then 
+                                analyzerService.Reanalyze(workspace, documentIds=documentIds, highPriority=true)
```
Also wondering if this is going to get us faster squiggles in the editor. Will probably move this change out of this PR though.

saul commented on vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs:
```diff
@@ -316,29 +328,50 @@ and
                 updated <- true
         
-        let updatedRefs = site.AssemblyReferences() |> hashSetIgnoreCase
+        (*
```
This will be put back - it's commented out as it's causing issues at the moment.

majocha replied:
I may be wrong but IIRC project updates must run on UI thread, contrary to what some comments in Roslyn say.

saul replied:
That's fine with this PR - the sources/flags notifier (which eventually calls through to this) is running on the UI thread.

saul commented on vsintegration/src/FSharp.ProjectSystem.Base/Project/ProjectNode.cs:
```diff
@@ -697,7 +697,7 @@ public override string Url
         public void SetDebugLogger(Microsoft.Build.Framework.ILogger debugLogger)
         {
-            myDebugLogger = debugLogger;
+            //myDebugLogger = debugLogger;
```
Temporary change, should not have committed

forki commented:
How could I miss that. Probably just read the email. 
Anyways: that's awesome! :kiss:

saul commented:
Yeah I added the perf stats in after I posted. I assume you've got the repo setup to email you immediately whenever there's action 😆

saul commented:
With regards to being able to export/get the ProjectFlagsService - any more ideas on this? Even if I use just [<Export>] and try to get the service via IComponentModel, it can never find it. I've no idea what I'm doing wrong - it seems to be exactly how other projects like PTVS and the GitHub VS extension use MEF.

cc @majocha

majocha commented:
@saul from what I gather ProjectPackage, where you want to import it,  is not MEF composed itself. To get the ProjectFlagsService there you would need to go through SComponentModel. I tried to get SComponentModel service and then import a small dummy part. It fails, so it's not just ProjectFlagsService.
Is it worth the trouble here? Do you plan on reusing ProjectFlagsService somewhere else?

saul commented:
@majocha yeah I was hoping to use it in FSharp.Editor. I want to call PrioritiseProject when a document is opened.

saul commented:
In the latest commit I've fixed #1944 and #2916 as I moved some LanguageService code over to use IVsRunningDocumentTable instead of hooking into SetupNewTextView (that's been removed entirely). I've also re-enabled the Roslyn workspace syncronisation that I disabled in the previous commit.

Also some more stats - the VisualFSharp solution now loads in 10 seconds (down from 20).

forki commented:
That's so awesome. Can we already merge it and just iterate on it?

Am 25.04.2017 6:24 nachm. schrieb "Saul Rennison" <notifications@github.com
>:

In the latest commit I've fixed #1944
<https://github.com/Microsoft/visualfsharp/issues/1944> and #2916
<https://github.com/Microsoft/visualfsharp/issues/2916> as I moved some
LanguageService code over to use IVsRunningDocumentTable instead of hooking
into SetupNewTextView (that's been removed entirely). I've also re-enabled
the Roslyn workspace syncronisation that I disabled in the previous commit.

Also some more stats - the VisualFSharp solution now loads in 10 seconds
(down from 20).

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/visualfsharp/pull/2909#issuecomment-297085498>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AADgNO1SIxFSDbiCY0zD_KRo4CECVhoSks5rzh46gaJpZM4NGYh5>
.

vasily-kirichenko commented:
FSharp.Compiler project cannot load with  this PR:

![image](https://cloud.githubusercontent.com/assets/873919/25398770/d0beff40-29f5-11e7-8927-df0bb054e540.png)

saul commented:
@vasily-kirichenko I believe that's because some of the references aren't of a legal form.

vasily-kirichenko commented:
@saul I tried to load FSharp.Configuration - it loaded insanely fast, but nothing works (coloring, etc.), no errors anywhere.

saul commented:
@vasily-kirichenko @dsyme I think I've found the cause of why you aren't getting semantic colourisation. (I do get block structure and syntactic). See the reactor log below:

```
Reactor: receiving..., remaining 0, mem 41, gc2 9
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 41, gc2 9
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 24.0012ms
Reactor: receiving..., remaining 1, mem 41, gc2 9
Reactor: --> set background op, remaining 0, mem 41, gc2 9
Reactor: receiving..., remaining 0, mem 41, gc2 9
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 41, gc2 9
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.0001ms
Reactor: receiving..., remaining 1, mem 41, gc2 9
Reactor: --> set background op, remaining 0, mem 41, gc2 9
Reactor: receiving..., remaining 0, mem 41, gc2 9
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 53, gc2 9
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.0001ms
Reactor: receiving..., remaining 1, mem 53, gc2 9
Reactor: --> set background op, remaining 0, mem 53, gc2 9
Reactor: receiving..., remaining 0, mem 53, gc2 9
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 57, gc2 9
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 1.9996ms
Reactor: receiving..., remaining 1, mem 57, gc2 9
Reactor: --> set background op, remaining 0, mem 57, gc2 9
Reactor: receiving..., remaining 0, mem 57, gc2 9
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 62, gc2 9
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.0005ms
Reactor: receiving..., remaining 1, mem 62, gc2 9
Reactor: --> set background op, remaining 0, mem 62, gc2 9
Reactor: receiving..., remaining 0, mem 62, gc2 9
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 62, gc2 9
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.9995ms
Reactor: receiving..., remaining 1, mem 62, gc2 9
Reactor: --> set background op, remaining 0, mem 62, gc2 9
Reactor: receiving..., remaining 0, mem 62, gc2 9
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 53, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 3.0013ms
Reactor: receiving..., remaining 1, mem 53, gc2 10
Reactor: --> set background op, remaining 0, mem 53, gc2 10
Reactor: receiving..., remaining 0, mem 53, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 53, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 1.9999ms
Reactor: receiving..., remaining 1, mem 53, gc2 10
Reactor: --> set background op, remaining 0, mem 53, gc2 10
Reactor: receiving..., remaining 0, mem 53, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 54, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 1.9978ms
Reactor: receiving..., remaining 1, mem 54, gc2 10
Reactor: --> set background op, remaining 0, mem 54, gc2 10
Reactor: receiving..., remaining 0, mem 54, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 55, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 1.9998ms
Reactor: receiving..., remaining 1, mem 55, gc2 10
Reactor: --> set background op, remaining 0, mem 55, gc2 10
Reactor: receiving..., remaining 0, mem 55, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 56, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 10.9998ms
Reactor: receiving..., remaining 1, mem 56, gc2 10
Reactor: --> set background op, remaining 0, mem 56, gc2 10
Reactor: receiving..., remaining 0, mem 56, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 57, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.001ms
Reactor: receiving..., remaining 1, mem 57, gc2 10
Reactor: --> set background op, remaining 0, mem 57, gc2 10
Reactor: receiving..., remaining 0, mem 57, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 58, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 3.0009ms
Reactor: receiving..., remaining 1, mem 58, gc2 10
Reactor: --> set background op, remaining 0, mem 58, gc2 10
Reactor: receiving..., remaining 0, mem 58, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 57, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 1.9967ms
Reactor: receiving..., remaining 1, mem 57, gc2 10
Reactor: --> set background op, remaining 0, mem 57, gc2 10
Reactor: receiving..., remaining 0, mem 57, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 55, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.0007ms
Reactor: receiving..., remaining 1, mem 55, gc2 10
Reactor: --> set background op, remaining 0, mem 55, gc2 10
Reactor: receiving..., remaining 0, mem 55, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 59, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.0011ms
Reactor: receiving..., remaining 1, mem 59, gc2 10
Reactor: --> set background op, remaining 0, mem 59, gc2 10
Reactor: receiving..., remaining 0, mem 59, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 60, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 1.9989ms
Reactor: receiving..., remaining 1, mem 60, gc2 10
Reactor: --> set background op, remaining 0, mem 60, gc2 10
Reactor: receiving..., remaining 0, mem 60, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 61, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 2.9996ms
Reactor: receiving..., remaining 1, mem 61, gc2 10
Reactor: --> set background op, remaining 0, mem 61, gc2 10
Reactor: receiving..., remaining 0, mem 61, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 61, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 3.0003ms
Reactor: receiving..., remaining 1, mem 61, gc2 10
Reactor: --> set background op, remaining 0, mem 61, gc2 10
Reactor: receiving..., remaining 0, mem 61, gc2 10
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: --> InvalidateConfiguration, remaining 0, mem 61, gc2 10
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 1.9977ms
Reactor: receiving..., remaining 1, mem 61, gc2 10
Reactor: --> set background op, remaining 0, mem 61, gc2 10
Reactor: receiving..., remaining 0, mem 61, gc2 10
Reactor: --> background step, remaining 0, mem 66, gc2 10
Reactor: <-- background step, remaining 0, took 4508.7542ms
Reactor: receiving..., remaining 0, mem 126, gc2 11
Reactor: --> background step, remaining 0, mem 122, gc2 11
Reactor: <-- background step, remaining 0, took 3.9976ms
Reactor: receiving..., remaining 0, mem 122, gc2 11
Reactor: --> background step, remaining 0, mem 122, gc2 11
Reactor: <-- background step, remaining 0, took 2.9987ms
Reactor: receiving..., remaining 0, mem 123, gc2 11
Reactor: --> background step, remaining 0, mem 123, gc2 11
Reactor: <-- background step, remaining 0, took 2.9999ms
Reactor: receiving..., remaining 0, mem 123, gc2 11
Reactor: --> background step, remaining 0, mem 123, gc2 11
Reactor: <-- background step, remaining 0, took 3.0008ms
Reactor: receiving..., remaining 0, mem 123, gc2 11
Reactor: --> background step, remaining 0, mem 123, gc2 11
Reactor: <-- background step, remaining 0, took 13.0257ms
Reactor: receiving..., remaining 0, mem 124, gc2 11
Reactor: --> background step, remaining 0, mem 124, gc2 11
Reactor: <-- background step, remaining 0, took 734.9841ms
Reactor: receiving..., remaining 0, mem 128, gc2 11
Reactor: --> background step, remaining 0, mem 128, gc2 11
Reactor: <-- background step, remaining 0, took 993.9752ms
Reactor: receiving..., remaining 0, mem 134, gc2 11
Reactor: --> background step, remaining 0, mem 134, gc2 11
Reactor: <-- background step, remaining 0, took 72.0216ms
Reactor: receiving..., remaining 0, mem 128, gc2 11
Reactor: --> background step, remaining 0, mem 128, gc2 11
Reactor: <-- background step, remaining 0, took 99.9971ms
Reactor: receiving..., remaining 0, mem 129, gc2 11
Reactor: --> background step, remaining 0, mem 129, gc2 11
Reactor: enqueue ParseFileInProject LanguageService.fs, length 0
Reactor: <-- background step, remaining 1, took 70.0207ms
Reactor: receiving..., remaining 1, mem 129, gc2 11
Reactor: --> ParseFileInProject LanguageService.fs, remaining 0, mem 129, gc2 11
Reactor: <-- ParseFileInProject LanguageService.fs, remaining 0, took 258.0112ms
Reactor: receiving..., remaining 0, mem 130, gc2 11
Reactor: enqueue ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 0
Reactor: --> ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 0, mem 105, gc2 12
Reactor: <-- ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 0, took 80.9957ms
Reactor: receiving..., remaining 0, mem 103, gc2 12
Reactor: --> background step, remaining 0, mem 110, gc2 12
Reactor: enqueue InvalidateConfiguration, length 0
Reactor: <-- background step, remaining 1, took 945.9585ms
Reactor: receiving..., remaining 1, mem 116, gc2 12
Reactor: --> InvalidateConfiguration, remaining 0, mem 116, gc2 12
Reactor: enqueue start background, length 0
Reactor: <-- InvalidateConfiguration, remaining 1, took 3.0001ms
Reactor: receiving..., remaining 1, mem 116, gc2 12
Reactor: --> set background op, remaining 0, mem 116, gc2 12
Reactor: receiving..., remaining 0, mem 116, gc2 12
Reactor: --> background step, remaining 0, mem 115, gc2 12
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 0
Reactor: enqueue InvalidateConfiguration, length 1
Reactor: enqueue MatchBraces LanguageService.fs, length 2
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 3
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 4
Reactor: enqueue InvalidateConfiguration, length 5
Reactor: enqueue MatchBraces LanguageService.fs, length 6
Reactor: enqueue ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 7
Reactor: enqueue ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 8
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 9
Reactor: enqueue InvalidateConfiguration, length 10
Reactor: enqueue InvalidateConfiguration, length 11
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 12
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 13
Reactor: enqueue InvalidateConfiguration, length 14
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 15
Reactor: enqueue InvalidateConfiguration, length 16
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 17
Reactor: enqueue InvalidateConfiguration, length 18
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 19
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 20
Reactor: enqueue InvalidateConfiguration, length 21
Reactor: enqueue MatchBraces LanguageService.fs, length 22
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 23
Reactor: enqueue ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 24
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 25
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 26
Reactor: enqueue InvalidateConfiguration, length 27
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 28
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 29
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 30
Reactor: enqueue InvalidateConfiguration, length 31
Reactor: enqueue InvalidateConfiguration, length 32
Reactor: enqueue InvalidateConfiguration, length 33
Reactor: enqueue InvalidateConfiguration, length 34
Reactor: enqueue InvalidateConfiguration, length 35
Reactor: enqueue InvalidateConfiguration, length 36
Reactor: enqueue InvalidateConfiguration, length 37
Reactor: enqueue InvalidateConfiguration, length 38
Reactor: enqueue InvalidateConfiguration, length 39
Reactor: enqueue InvalidateConfiguration, length 40
Reactor: <-- background step, remaining 41, took 324170.6876ms
Reactor: receiving..., remaining 41, mem 254, gc2 29
Reactor: receiving..., remaining 40, mem 254, gc2 29
Reactor: --> InvalidateConfiguration, remaining 39, mem 254, gc2 29
Reactor: enqueue start background, length 39
Reactor: <-- InvalidateConfiguration, remaining 40, took 0.9996ms
Reactor: receiving..., remaining 40, mem 254, gc2 29
Reactor: receiving..., remaining 39, mem 254, gc2 29
Reactor: receiving..., remaining 38, mem 254, gc2 29
Reactor: receiving..., remaining 37, mem 254, gc2 29
Reactor: --> InvalidateConfiguration, remaining 36, mem 254, gc2 29
Reactor: enqueue start background, length 36
Reactor: <-- InvalidateConfiguration, remaining 37, took 1.9995ms
Reactor: receiving..., remaining 37, mem 254, gc2 29
Reactor: receiving..., remaining 36, mem 254, gc2 29
Reactor: --> ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 35, mem 254, gc2 29
Reactor: <-- ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 35, took 85086.3554ms
Reactor: receiving..., remaining 35, mem 193, gc2 31
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 35
Reactor: receiving..., remaining 34, mem 193, gc2 31
Reactor: receiving..., remaining 34, mem 193, gc2 31
Reactor: --> InvalidateConfiguration, remaining 33, mem 193, gc2 31
Reactor: enqueue start background, length 33
Reactor: <-- InvalidateConfiguration, remaining 34, took 1ms
Reactor: receiving..., remaining 34, mem 193, gc2 31
Reactor: --> InvalidateConfiguration, remaining 33, mem 193, gc2 31
Reactor: enqueue start background, length 33
Reactor: <-- InvalidateConfiguration, remaining 34, took 0.9997ms
Reactor: receiving..., remaining 34, mem 193, gc2 31
Reactor: receiving..., remaining 33, mem 193, gc2 31
Reactor: receiving..., remaining 32, mem 193, gc2 31
Reactor: --> InvalidateConfiguration, remaining 31, mem 193, gc2 31
Reactor: enqueue start background, length 31
Reactor: <-- InvalidateConfiguration, remaining 32, took 1ms
Reactor: receiving..., remaining 32, mem 193, gc2 31
Reactor: receiving..., remaining 31, mem 193, gc2 31
Reactor: --> InvalidateConfiguration, remaining 30, mem 193, gc2 31
Reactor: enqueue start background, length 30
Reactor: <-- InvalidateConfiguration, remaining 31, took 0.9998ms
Reactor: receiving..., remaining 31, mem 193, gc2 31
Reactor: receiving..., remaining 30, mem 193, gc2 31
Reactor: --> InvalidateConfiguration, remaining 29, mem 193, gc2 31
Reactor: enqueue start background, length 29
Reactor: <-- InvalidateConfiguration, remaining 30, took 1.0001ms
Reactor: receiving..., remaining 30, mem 193, gc2 31
Reactor: receiving..., remaining 29, mem 193, gc2 31
Reactor: receiving..., remaining 28, mem 193, gc2 31
Reactor: --> InvalidateConfiguration, remaining 27, mem 193, gc2 31
Reactor: enqueue start background, length 27
Reactor: <-- InvalidateConfiguration, remaining 28, took 0.9994ms
Reactor: receiving..., remaining 28, mem 193, gc2 31
Reactor: --> MatchBraces LanguageService.fs, remaining 27, mem 193, gc2 31
Reactor: <-- MatchBraces LanguageService.fs, remaining 27, took 26.0197ms
Reactor: receiving..., remaining 27, mem 196, gc2 31
Reactor: receiving..., remaining 26, mem 196, gc2 31
Reactor: --> ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 25, mem 196, gc2 31
Reactor: <-- ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 25, took 0ms
Reactor: receiving..., remaining 25, mem 196, gc2 31
Reactor: receiving..., remaining 24, mem 196, gc2 31
Reactor: receiving..., remaining 23, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 22, mem 196, gc2 31
Reactor: enqueue start background, length 22
Reactor: <-- InvalidateConfiguration, remaining 23, took 0.9771ms
Reactor: receiving..., remaining 23, mem 196, gc2 31
Reactor: receiving..., remaining 22, mem 196, gc2 31
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 21, mem 196, gc2 31
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 21, took 1.0008ms
Reactor: receiving..., remaining 21, mem 196, gc2 31
Reactor: receiving..., remaining 20, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 19, mem 196, gc2 31
Reactor: enqueue start background, length 19
Reactor: <-- InvalidateConfiguration, remaining 20, took 2.9995ms
Reactor: receiving..., remaining 20, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 19, mem 196, gc2 31
Reactor: enqueue start background, length 19
Reactor: <-- InvalidateConfiguration, remaining 20, took 1.9992ms
Reactor: receiving..., remaining 20, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 19, mem 196, gc2 31
Reactor: enqueue start background, length 19
Reactor: <-- InvalidateConfiguration, remaining 20, took 1.9965ms
Reactor: receiving..., remaining 20, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 19, mem 196, gc2 31
Reactor: enqueue start background, length 19
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 19
Reactor: <-- InvalidateConfiguration, remaining 20, took 22.0241ms
Reactor: receiving..., remaining 21, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 20, mem 196, gc2 31
Reactor: enqueue start background, length 20
Reactor: <-- InvalidateConfiguration, remaining 21, took 22.0134ms
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 21
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 21
Reactor: receiving..., remaining 21, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 22, mem 196, gc2 31
Reactor: enqueue start background, length 22
Reactor: <-- InvalidateConfiguration, remaining 23, took 1.9952ms
Reactor: receiving..., remaining 23, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 22, mem 196, gc2 31
Reactor: enqueue start background, length 22
Reactor: <-- InvalidateConfiguration, remaining 23, took 2.0826ms
Reactor: receiving..., remaining 23, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 22, mem 196, gc2 31
Reactor: enqueue start background, length 22
Reactor: <-- InvalidateConfiguration, remaining 23, took 2.0075ms
Reactor: receiving..., remaining 23, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 22, mem 196, gc2 31
Reactor: enqueue start background, length 22
Reactor: <-- InvalidateConfiguration, remaining 23, took 2.0039ms
Reactor: receiving..., remaining 23, mem 196, gc2 31
Reactor: --> InvalidateConfiguration, remaining 22, mem 196, gc2 31
Reactor: enqueue start background, length 22
Reactor: <-- InvalidateConfiguration, remaining 23, took 1.9664ms
Reactor: receiving..., remaining 23, mem 196, gc2 31
Reactor: --> set background op, remaining 22, mem 196, gc2 31
Reactor: receiving..., remaining 22, mem 196, gc2 31
Reactor: --> set background op, remaining 21, mem 196, gc2 31
Reactor: receiving..., remaining 21, mem 196, gc2 31
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 20, mem 196, gc2 31
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 20, took 0ms
Reactor: receiving..., remaining 20, mem 196, gc2 31
Reactor: --> set background op, remaining 19, mem 196, gc2 31
Reactor: receiving..., remaining 19, mem 196, gc2 31
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 20
Reactor: --> set background op, remaining 18, mem 196, gc2 31
Reactor: receiving..., remaining 19, mem 196, gc2 31
Reactor: --> set background op, remaining 18, mem 196, gc2 31
Reactor: receiving..., remaining 18, mem 196, gc2 31
Reactor: --> set background op, remaining 17, mem 196, gc2 31
Reactor: receiving..., remaining 17, mem 196, gc2 31
Reactor: --> set background op, remaining 16, mem 196, gc2 31
Reactor: receiving..., remaining 16, mem 196, gc2 31
Reactor: --> set background op, remaining 15, mem 196, gc2 31
Reactor: receiving..., remaining 15, mem 196, gc2 31
Reactor: --> set background op, remaining 14, mem 196, gc2 31
Reactor: receiving..., remaining 14, mem 196, gc2 31
Reactor: --> set background op, remaining 13, mem 196, gc2 31
Reactor: receiving..., remaining 13, mem 196, gc2 31
Reactor: --> set background op, remaining 12, mem 196, gc2 31
Reactor: receiving..., remaining 12, mem 196, gc2 31
Reactor: --> set background op, remaining 11, mem 196, gc2 31
Reactor: receiving..., remaining 11, mem 196, gc2 31
Reactor: --> set background op, remaining 10, mem 196, gc2 31
Reactor: receiving..., remaining 10, mem 196, gc2 31
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 9, mem 196, gc2 31
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 9
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 10
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 11
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 12, took 61422.0086ms
Reactor: receiving..., remaining 12, mem 190, gc2 32
Reactor: --> set background op, remaining 11, mem 190, gc2 32
Reactor: receiving..., remaining 11, mem 190, gc2 32
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 10, mem 190, gc2 32
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 10, took 93.0119ms
Reactor: receiving..., remaining 10, mem 219, gc2 32
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 10
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 9, mem 219, gc2 32
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 10, took 0ms
Reactor: receiving..., remaining 10, mem 219, gc2 32
Reactor: --> set background op, remaining 9, mem 219, gc2 32
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 10
Reactor: receiving..., remaining 9, mem 219, gc2 32
Reactor: --> set background op, remaining 9, mem 219, gc2 32
Reactor: receiving..., remaining 9, mem 219, gc2 32
Reactor: --> set background op, remaining 8, mem 219, gc2 32
Reactor: receiving..., remaining 8, mem 219, gc2 32
Reactor: --> set background op, remaining 7, mem 219, gc2 32
Reactor: receiving..., remaining 7, mem 219, gc2 32
Reactor: --> set background op, remaining 6, mem 219, gc2 32
Reactor: receiving..., remaining 6, mem 219, gc2 32
Reactor: receiving..., remaining 5, mem 219, gc2 32
Reactor: receiving..., remaining 4, mem 219, gc2 32
Reactor: receiving..., remaining 3, mem 219, gc2 32
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 2, mem 219, gc2 32
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 2, took 1.9999ms
Reactor: receiving..., remaining 2, mem 220, gc2 32
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 2
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 1
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 1, mem 220, gc2 32
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 3, took 16677.1243ms
Reactor: receiving..., remaining 3, mem 253, gc2 33
Reactor: receiving..., remaining 2, mem 253, gc2 33
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 1, mem 253, gc2 33
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 1
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 2, took 600.9489ms
Reactor: receiving..., remaining 2, mem 250, gc2 33
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 1, mem 250, gc2 33
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 1, took 113.9929ms
Reactor: receiving..., remaining 1, mem 224, gc2 33
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 0, mem 224, gc2 33
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 1
Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 0, took 0ms
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 0
Reactor: receiving..., remaining 1, mem 224, gc2 33
Reactor: --> ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 1, mem 224, gc2 33
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 1
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 2
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 3
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 4
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 5
Reactor: enqueue ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, length 6
```

Search that log for long running tasks. Especially noteworthy:

- Reactor: <-- background step, remaining 41, took 324170.6876ms
- Reactor: <-- ParseFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 35, took 85086.3554ms
- Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 12, took 61422.0086ms
- Reactor: <-- ParseAndCheckFileInProject C:\Code\visualfsharp\vsintegration\src\FSharp.Editor\LanguageService\LanguageService.fs, remaining 3, took 16677.1243ms

I've no idea why these are taking so long... time to break out the profiler? 😦

dsyme commented on vsintegration/src/FSharp.ProjectSystem.FSharp/ProjectFlagsService.fs:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        async {
+            while true do
```
This basically looks very clear.  But I'd somehow like to see some safeguards where we try/catch any exceptions in chunks of this code.  Infinite loops starting mailbox processors make me nervous if exceptions might happen.  

And are we sure the solution events get raised reliably?  

Basically I'm wondering "what can go catastrophically wrong that can cause this service to malfunction in really bad ways"

saul replied:
I'm confident that the solution events get raised reliably, yes.

With regards to try/catches, where abouts would you like those? It's worth noting that the project system DoMSBuildSubmission is already very resilient to exceptions and will always call our callback on the UI thread.

dsyme commented on vsintegration/src/FSharp.ProjectSystem.FSharp/ProjectFlagsService.fs:
```diff
@@ -0,0 +1,128 @@
+            
+            // Keep trying to grab the exclusive lock for a design time build
+            while accessor.BeginDesignTimeBuild() |> ErrorHandler.Failed do
+                do! Async.Sleep 2000
```
2sec is a long pause - does it need to be so long?

saul replied:
No I'll reduce this to less than a second.

dsyme commented on vsintegration/src/FSharp.ProjectSystem.FSharp/ProjectFlagsService.fs:
```diff
@@ -0,0 +1,128 @@
+                            // Having an exclusive lock on design-time builds means the user
+                            // can't initiate any builds themselves. Wait a few seconds before
+                            // trying to grab one again.
+                            do! Async.Sleep 2000
```
Likewise - why such a comparatively long pause?

saul replied:
It leaves some time so that the user can initiate builds/move files around in the Solution Explorer. When a design-time build is running, quite a few parts of the UI are locked out.

dsyme commented on vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs:
```diff
@@ -121,26 +127,38 @@ type internal ProjectSitesAndFiles() =
                 | _ -> None)
         | None -> Seq.empty
                     
+    static let projectCache = ConcurrentDictionary<string,FSharpProjectOptions>()
```
What removes entries from this cache?

saul replied:
Nothing yet - good catch

saul replied:
Fixed

dsyme commented:
@saul Please take a look at https://github.com/Microsoft/visualfsharp/pull/3002 for a different fix.

If we went with that fix would you want to keep this? It could still be valuable since you mention other bug fixes?

dsyme commented on vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs:
```diff
@@ -121,26 +127,41 @@ type internal ProjectSitesAndFiles() =
-            UnresolvedReferences = None
-            OriginalLoadReferences = []
-            ExtraProjectInfo=extraProjectInfo }   
+    and getProjectOptionsForProjectSite(projectSite:IProjectSite, fileName:string, extraProjectInfo, serviceProvider, cacheUsage:CacheUsage) =
```
This will clash with my other proposed fix.  The project options are already stored in a table in LanguageService.fs, it seems to me we should just callback to there to look up the table.

dsyme commented on vsintegration/src/FSharp.ProjectSystem.Base/Project/AssemblyReferenceNode.cs:
```diff
@@ -76,7 +76,7 @@ public override string Url
             get
             {
-                return this.myAssemblyPath;
+                return Path.GetFullPath(this.myAssemblyPath);
```
This fix still looks useful?

saul replied:
This was used by the caching code to determine if two references were the same. `this.myAssemblyPath` is not absolute and so it made comparing them difficult when the rooted-ness was inconsistent within a solution. It's not a bug fix per se.

dsyme commented on vsintegration/src/FSharp.ProjectSystem.Base/Project/ProjectReferenceNode.cs:
```diff
@@ -872,11 +872,12 @@ private static System.Runtime.Versioning.FrameworkName GetProjectTargetFramework
-
-            string targetFrameworkMoniker = (string)otherTargetFrameworkMonikerObj;
+            object otargetFrameworkMoniker = null;
+            int hr = hierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID4.VSHPROPID_TargetFrameworkMoniker, out otargetFrameworkMoniker);
```
Is this the correctness fix?

saul replied:
No this stops an occasional assertion on startup

dsyme replied:
ok, thanks

saul commented on vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs:
```diff
@@ -400,24 +494,13 @@ and
-            | _ -> ()
+    
+    member this.TrackDocument(fileName) =
+        match VsRunningDocumentTable.FindDocumentWithoutLocking(package.RunningDocumentTable, fileName) with
```
@dsyme this file essentially encapsulates the correctness fixes. However with 15.3 now in preview we can move over to the Miscellaneous Files workspace. Read the comments on https://github.com/dotnet/roslyn/pull/17263 and my comment about Miscellaneous Files in this file.

saul commented:
@dsyme I'll port over the RunningDocumentTable changes that fix #2916 and #1944 to a smaller PR in due course 😄

dsyme commented:
> @dsyme I'll port over the RunningDocumentTable changes that fix #2916 and #1944 to a smaller PR in due course 😄

That would be fantastic, thanks!

dsyme commented:
@saul This PR is now really old, and I will close it.  Can you reopen or resubmit if you want to resurect it? thanks
